app/javascript/packages/spinner-button/spinner-button-element.ts

```diff
@@ -25,38 +47,29 @@ export class SpinnerButtonElement extends HTMLElement {
     return Number(this.getAttribute('long-wait-duration-ms')) || DEFAULT_LONG_WAIT_DURATION_MS;
   }
 
-  connectedCallback() {
-    this.elements = {
-      button: this.querySelector('a,button:not([type]),[type="submit"],[type="button"]')!,
-      actionMessage: this.querySelector('.spinner-button__action-message')!,
-    };
-
-    if (this.spinOnClick) {
-      this.elements.button.addEventListener('click', () => this.toggleSpinner(true));
+  bindSpinOnClick() {
+    if (!this.spinOnClick) {
+      return;
     }
-    this.addEventListener('spinner.start', () => this.toggleSpinner(true));
-    this.addEventListener('spinner.stop', () => this.toggleSpinner(false));
-  }
 
-  disconnectedCallback() {
-    window.clearTimeout(this.#longWaitTimeout);
+    if (this.form) {
+      this.form.addEventListener('submit', () => this.toggleSpinner(true));
+    } else {
+      this.button.addEventListener('click', () => this.toggleSpinner(true));
+    }
   }
 
   toggleSpinner(isVisible: boolean) {
-    const { button, actionMessage } = this.elements;
     this.classList.toggle('spinner-button--spinner-active', isVisible);
 
-    // Avoid setting disabled immediately to allow click event to propagate for form submission.
-    setTimeout(() => {
-      if (isVisible) {
-        button.setAttribute('disabled', '');
-      } else {
-        button.removeAttribute('disabled');
-      }
-    }, 0);
-
-    if (actionMessage) {
-      actionMessage.textContent = isVisible ? (actionMessage.dataset.message as string) : '';
+    if (isVisible) {
+      this.button.setAttribute('disabled', '');
+    } else {
+      this.button.removeAttribute('disabled');
+    }
```

Comment:
If you make `toggleSpinner` an observedAttribute / callback you wouldn't need to set the attributes here -- they could just be reactive. Not sure if there's any real advantage thought because it might be more verbose in the end.

Reply:
Are you thinking something like a `spinner-active` attribute for the element that we'd reactively disable/enable the button, in place of what we currently use the `.spinner-button--spinner-active` CSS class for? I quite like that.

Reply:
Yeah, though on second though might be cleaner to keep the `toggleSpinner`'s setting of the button disabled attributes, like so:
```javascript
static get observedAttributes(){
    return [
      'spinner-active'
    ];
  }

  attributeChangedCallback(name, oldVal, newVal){
    if(name === 'spinner-active'){
      this.toggleSpinner(!!newVal);
    }
  }

  toggleSpinner(isActive){
    if(!isActive){
      this.button.removeAttribute("disabled");
    } else {
      this.button.setAttribute("disabled", "");
    }
  }
```
  
And now you can just use `spinner-button[spinner-active]` as the selector for animation, rather than a class

Reply:
Or better yet:
```javascript
attributeChangedCallback(name, oldVal, newVal){
    if(name === 'spinner-active'){
      this.button.toggleAttribute('disabled', !!newVal);
    }
  }
```

Reply:
I like this plan. I'd probably want the separate method since we'll still need to deal with the delayed display of the "action message". I might propose to separate out that refactoring to a quick follow-on pull request, to keep this one a bit more focused on submit handling.